lib/cdo/aws/dms.rb

```diff
@@ -2,6 +2,13 @@
 
 module Cdo
   class DMS
+    REDSHIFT_SCHEMAS_IMPORTED_FROM_DATABASE = %w(
+      dashboard_production
+      dashboard_production_pii
+      pegasus
+      pegasus_pii
+    ).freeze
```

Comment:
This breaks the abstraction of this module, requiring us to update both here and in the tasks YML when we change target schemas. Can we derive these values from the content YML to maintain the existing separation between code and content?

Reply:
I'd prefer that as well, but I haven't found a way to derive a distinct list of Redshift schemas from `tasks.yml` nor from the DMS Mapping Rules we generate from `tasks.yml`.  DMS employs input-centric transformation rules ("what mapping/transformation should I carry out with this set of input/source tables?").  To programmatically identify the Redshift schemas to which our Replication Tasks export data, I think we would need to re-implement the DMS mapping rules engine.  For example, we'd probably need to interpret [a `selection` rule](https://github.com/code-dot-org/code-dot-org/blob/8adf029f2ff2e293aef65ad5a456f81005ad5a3e/lib/cdo/aws/dms.rb#L18-L25) for the MySQL `dashboard.census_submissions` table and a[ subsequent `transformation` rule](https://github.com/code-dot-org/code-dot-org/blob/8adf029f2ff2e293aef65ad5a456f81005ad5a3e/lib/cdo/aws/dms.rb#L62-L71) that suffixes the output Redshift schema with `_pii` to derive that one of the Redshift schemas is `dashboard_production_pii`.

I'd appreciate advice on ways to get the list of Redshift schemas to which DMS exports tables.  The compromise we have currently is to declare this constant next to the method that generates the DMS Rules.  The `DMS.tasks` method reliance on `CDO.#{schema}_db_name` means that any new Redshift schemas we export to would require updating this method as well, and the person making that change is likely to notice that the constant needs to be updated as well.

Reply:
Ah I see, since `CDO["#{schema}_db_name"]` is in the module code to derive the environment-specific schema name, and the transformation possibly appends `_pii` to the schema name as well.

Here's two approaches I can come up with to at least keep the envrionment-config specific names out of this module:

1. scan the generated task-definitions for `"schema-name":"x"` keys, then combine with `_pii`:
```ruby
Cdo::DMS.tasks(aws_dir('dms/tasks.yml')).to_json.scan(/"schema-name":"(\w+)"/).uniq.product(['', '_pii']).map(&:join)
=> ["dashboard_production", "dashboard_production_pii", "pegasus", "pegasus_pii"]
```

2. specify 'dashboard' and 'pegasus' directly to get the db names from config, then combine with `_pii`:
```ruby
%w(dashboard pegasus).map {|schema| CDO["#{schema}_db_name"]}.product(['', '_pii']).map(&:join)
=> ["dashboard_production", "dashboard_production_pii", "pegasus", "pegasus_pii"]
```

Neither is great, but I prefer either slightly to hard-coding the `production`-specific db names in here.

Reply:
Thanks.  Very useful.  I went with the second option.